File: 1/exercise2/savethecat.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cstdio>
#include <queue>
#include <algorithm>

#define MAX_N 1005
#define MAX_M 1005

using namespace std;

string map[MAX_N];
int flooded[MAX_N][MAX_M];//flooded[i][j] = time in which [i,j] will be flooded
int catMoves[MAX_N][MAX_M]; //catMoves[i][j] = time in which the cat can reach [i,j]
vector< pair<int,int> > waterPumps,neighbors;
pair<int,int> father[MAX_N][MAX_M]; // previous coord in cat's shortest path to [i,j]
pair<int,int> cat;
int M,N;

bool operator > (pair<int,int> a, pair<int,int> b){
  if(a.first > b.first) return true;
  if(a.first == b.first && a.second > b.second) return true;
  return false;
}

string findMove(int x1,int y1, int x2, int y2){
  if( y1 == y2 ){
    if(x2==x1+1){
      return "D";
    }
    return "U";
  }
  else{
    if( y2 == y1 + 1){
      return "R";
    }
    return "L";
  }
}

string getPath(pair<int,int> endVertex){
  int currX=endVertex.first,currY=endVertex.second;
  pair<int,int> previous;
  string result;

  while(73){
    previous = father[currX][currY];
    if( previous == make_pair(-1,-1) ) break;
    result = result + findMove(previous.first,previous.second,currX,currY);
    currX = previous.first;
    currY = previous.second;
  }

  reverse(result.begin(),result.end());
  sort(result.begin(),result.end());

  return result;
}

vector< pair<int,int> > findNeighbors(pair<int,int> u){
  vector< pair<int,int> > result;
  int x = u.first, y = u.second;

  if(x-1 >= 0 && map[x-1][y] != 'X'){ // U
    result.push_back(make_pair(x-1,y));
  }

  if(x+1 <= N-1 && map[x+1][y] != 'X'){ // D
    result.push_back(make_pair(x+1,y));
  }

  if(y-1 >= 0 && map[x][y-1] != 'X'){ // L
    result.push_back(make_pair(x,y-1));
  }

  if(y+1 <= M-1 && map[x][y+1] != 'X'){ // R
    result.push_back(make_pair(x,y+1));
  }

  return result;
}


int main(int argc, char *argv[]) {

  ifstream inputFile;
  inputFile.open(argv[1]);

  int lines=0;
  while(getline(inputFile,map[lines++]));
  lines--;

  N = lines;
  M = map[0].size();

  for(int i=0; i<N; i++){
    for(int j=0; j<M; j++){
      if(map[i][j]=='W'){
        waterPumps.push_back(make_pair(i,j));
      }
      else if(map[i][j]=='A'){
        cat.first = i;
        cat.second = j;
      }
    }
  }

  // init flooded matrix
  for(int i=0; i<N; i++){
    for(int j=0; j<M; j++){
      flooded[i][j]=-1;
      catMoves[i][j]=-1;
    }
  }

  // bfs from water pumps

  queue< pair<int,int> > q;
  for( int i=0; i<waterPumps.size(); i++){
    q.push(waterPumps[i]);
    flooded[waterPumps[i].first][waterPumps[i].second] = 0;
  }

  while(!q.empty()){
    pair<int,int> v = q.front();
    q.pop();

    neighbors = findNeighbors(v);

    for(int i=0; i<neighbors.size(); i++){
      int x = neighbors[i].first, y = neighbors[i].second;
      if(flooded[x][y] == -1){
        flooded[x][y] = flooded[v.first][v.second] + 1;
        q.push(neighbors[i]);
      }
    }
  }

  // BFS for cat moves - save distances and fathers
  q.push(cat);
  catMoves[cat.first][cat.second] = 0;
  father[cat.first][cat.second] = make_pair(-1,-1);

  while(!q.empty()){
    pair<int,int> v = q.front();
    q.pop();

    neighbors = findNeighbors(v);
    for(int i=0; i<neighbors.size(); i++){
      int x = neighbors[i].first, y = neighbors[i].second;
      if(catMoves[x][y]==-1){
        catMoves[x][y] = catMoves[v.first][v.second] + 1;
        q.push(neighbors[i]);
        father[x][y] = v;
      }
    }
  }

  pair<int,int> finalPosition = make_pair(N,M);
  int time=-1;
  bool catIsSafe = false;

  for(int i=0;i<N;i++){
    for(int j=0;j<M;j++){
      if(catMoves[i][j]==-1) continue;

      if(flooded[i][j]==-1 && map[i][j]!='X'){
        catIsSafe = true;
        if(finalPosition > make_pair(i,j)) finalPosition = make_pair(i,j);
      }
      else if(!catIsSafe && flooded[i][j]>=time && catMoves[i][j]<flooded[i][j]){
        if( time < flooded[i][j] || finalPosition > make_pair(i,j) ){
          finalPosition = make_pair(i,j);
          time = flooded[i][j];
        }
      }
    }
  }


  if(catIsSafe){
    cout<<"infinity\n";
  }
  if(finalPosition == cat){
    cout<<"stay\n";
  }
  else{
    string path = getPath(finalPosition);
    if(!catIsSafe) cout<<path.size()<<endl;
    cout<<path<<endl;
  }

  // for(int i=0;i<waterPumps.size();i++){
  //   printf("%d %d\n",waterPumps[i].first,waterPumps[i].second);
  // }


  // for(int i=0;i<N;i++){
  //   for(int j=0;j<M;j++){
  //     printf("%d ",flooded[i][j]);
  //   }
  //   printf("\n");
  // }
  //
  // printf("\nCat moves:\n");
  //
  // for(int i=0;i<N;i++){
  //   for(int j=0;j<M;j++){
  //     printf("%d ",catMoves[i][j]);
  //   }
  //   printf("\n");
  // }

}

File: savethecat/savethecat.py
from collections import deque
import sys

inputFile = open(sys.argv[1])
map = inputFile.read().split('\n')[:-1]
N = len(map)
M = len(map[0])

waterPumps = []
flooded = [[-1 for j in range(M)] for i in range(N)]
catMoves = [[-1 for j in range(M)] for i in range(N)]
parent = [[[-1,-1] for j in range(M)] for i in range(N)]


for i in range(N):
    for j in range(M):
        if map[i][j] == 'W':
            waterPumps.append([i,j])
            flooded[i][j] = 0

        elif map[i][j] == 'A':
            cat = [i,j]
            catMoves[i][j] = 0

def findNeighbors(v):
    result = []
    x = v[0]
    y = v[1]

    # D
    if x+1 <= N-1 and map[x+1][y]!='X':
        result.append([x+1,y])

    # L
    if y-1 >= 0 and map[x][y-1]!='X':
        result.append([x,y-1])

    # R
    if y+1 <= M-1 and map[x][y+1]!='X':
        result.append([x,y+1])

    # U
    if x-1 >= 0 and map[x-1][y]!='X':
        result.append([x-1,y])

    return result

def isPositionGreater(a,b):
    # returns a > b
    if a[0] > b[0]:
        return True
    elif a[0] == b[0] and a[1] > b[1]:
        return True
    return False

def findMove(x1,y1,x2,y2):
    if y1==y2:
        if x2==x1+1:
            return "D"
        return "U"
    else:
        if y2==y1+1:
            return "R"
        return "L"

def getPath(finalPos):
    currVertex = finalPos;
    previous = [-1,-1]
    result = []

    while(True):
        previous = parent[currVertex[0]][currVertex[1]]
        if previous == [-1,-1]:
            break
        result.append(findMove(previous[0],previous[1],currVertex[0],currVertex[1]))
        currVertex[0] = previous[0]
        currVertex[1] = previous[1]

    result.reverse()

    return ''.join(result)


def print2DMap(a):
    for i in range(len(a)):
        for j in range(len(a[0])):
            print(a[i][j],end=" ")
        print("")

# Water BFS
Q = deque(waterPumps)

while(len(Q)!=0):
    front = Q.popleft()
    floodingTime = flooded[front[0]][front[1]]
    neighbors = findNeighbors(front)

    for v in neighbors:
        if flooded[v[0]][v[1]] == -1:
            flooded[v[0]][v[1]] = floodingTime + 1
            Q.append(v)

# Cat BFS
Q = deque([cat])

catIsSage = False
time = -1
finalPosition = [N,M]

while(len(Q)!=0):
    u = Q.popleft()
    movingTime = catMoves[u[0]][u[1]]
    neighbors = findNeighbors(u)

    if flooded[u[0]][u[1]]!=-1 and movingTime >= flooded[u[0]][u[1]]:
        continue

    # Update best position:
    i = u[0]
    j = u[1]

    if flooded[i][j]==-1:
        catIsSage=True
        if isPositionGreater(finalPosition,[i,j]):
            finalPosition = [i,j]
    elif (not catIsSage) and (flooded[i][j] >= time) and (movingTime < flooded[i][j]):
        if time < flooded[i][j] or isPositionGreater(finalPosition,[i,j]):
            finalPosition = [i,j]
            time = flooded[i][j]

    # Push neighbors in queue
    for v in neighbors:
        if catMoves[v[0]][v[1]] == -1:
            catMoves[v[0]][v[1]] = movingTime + 1
            parent[v[0]][v[1]] = u
            Q.append(v)

# catIsSage = False
# time = -1
# finalPosition = [N,M]

# Find best position
# for i in range(N):
#     for j in range(M):
#         if catMoves[i][j]==-1:
#             continue
#
#         if flooded[i][j]==-1:
#             catIsSage=True
#             if isPositionGreater(finalPosition,[i,j]):
#                 finalPosition = [i,j]
#         elif (not catIsSage) and (flooded[i][j] >= time) and (catMoves[i][j] < flooded[i][j]):
#             if time < flooded[i][j] or isPositionGreater(finalPosition,[i,j]):
#                 finalPosition = [i,j]
#                 time = flooded[i][j]

if catIsSage:
    print("infinity")
else:
    print(time-1)

if finalPosition == cat:
    print("stay")
else:
    print(getPath(finalPosition))
